packages/flutter/lib/src/widgets/scroll_position.dart

```diff
@@ -489,11 +489,44 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       assert(minScrollExtent != null);
       assert(maxScrollExtent != null);
       assert(minScrollExtent <= maxScrollExtent);
+      final ScrollMetrics oldPosition = haveDimensions ? copyWith() : null;
       _minScrollExtent = minScrollExtent;
       _maxScrollExtent = maxScrollExtent;
+      final ScrollMetrics newPosition = haveDimensions ? copyWith() : null;
+      _didChangeViewportDimensionOrReceiveCorrection = false;
+      if (haveDimensions && !approveNewDimensions(oldPosition, newPosition))
+        return false;
       _haveDimensions = true;
       applyNewDimensions();
-      _didChangeViewportDimensionOrReceiveCorrection = false;
+    }
+    assert(!_didChangeViewportDimensionOrReceiveCorrection, 'Use approveNewDimensions() (and return true) to change the scroll offset during applyContentDimensions().');
+    return true;
+  }
+
+  /// Verifies that the new content and viewport dimensions are acceptable.
+  ///
+  /// Called by [applyContentDimensions] to determine its return value.
+  ///
+  /// Should return true if the current scroll offset is correct given
+  /// the new content and viewport dimensions.
+  ///
+  /// Should return false if it calls [correctPixels] to correct the
+  /// scroll offset given the new dimensions.
+  ///
+  /// This is only called when [haveDimensions] is true.
+  ///
+  /// The default implementation defers to [ScrollPhysics.adjustPositionForNewDimensions].
+  @protected
+  bool approveNewDimensions(ScrollMetrics oldPosition, ScrollMetrics newPosition) {
```

Comment:
A function called `approveNewDimensions` and returns a bool presents itself like a query but this method has side effects.  This is why I removed the side effects from this method and renamed it `doesApproveNewDimensions`.

Reply:
Having the method be separate makes it easier to override in custom scroll positions (much easier, given the privates accessed in the caller). But I agree about the name. I'll see if I can come up with a better one.

Reply:
Maybe just coming up with a different verb is enough, since `approve` usually is a boolean query, "do you approve?"

Reply:
how about `correctForNewDimensions`? (i've done that for now)